polaris-service/src/main/java/org/apache/polaris/service/persistence/MetadataCacheManager.java

```diff
@@ -0,0 +1,172 @@
+              .setContent(TableMetadataParser.toJson(metadata))
+              .build();
+      try {
+        return entityManager
+            .getMetaStoreManager()
+            .createEntityIfNotExists(
+                callContext,
+                PolarisEntity.toCoreList(resolvedEntities.getRawFullPath()),
+                tableMetadataEntity);
+      } catch (RuntimeException e) {
+        // PersistenceException (& other extension-specific exceptions) may not be in scope,
+        // but we can make a best-effort attempt to swallow it and just forego caching
+        if (e.toString().contains("PersistenceException")) {
+          return new PolarisMetaStoreManager.EntityResult(
+              PolarisMetaStoreManager.ReturnStatus.UNEXPECTED_ERROR_SIGNALED, e.getMessage());
+        } else {
+          throw e;
+        }
+      }
+    }
+  }
+
+  /** Return the cached {@link Table} entity, if one exists */
+  private static @NotNull Optional<TableMetadata> loadCachedTableMetadata(
+      TableIdentifier tableIdentifier,
+      PolarisCallContext callContext,
+      PolarisEntityManager entityManager,
+      PolarisResolutionManifestCatalogView resolvedEntityView) {
+    PolarisResolvedPathWrapper resolvedEntities =
+        resolvedEntityView.getPassthroughResolvedPath(tableIdentifier, PolarisEntitySubType.TABLE);
+    if (resolvedEntities == null) {
+      return Optional.empty();
+    } else {
+      TableLikeEntity entity = TableLikeEntity.of(resolvedEntities.getRawLeafEntity());
+      String metadataLocation = entity.getMetadataLocation();
+      PolarisMetaStoreManager.ListEntitiesResult metadataResult =
+          entityManager
+              .getMetaStoreManager()
+              .listEntities(
+                  callContext,
+                  PolarisEntity.toCoreList(resolvedEntities.getRawFullPath()),
+                  PolarisEntityType.TABLE_METADATA,
+                  PolarisEntitySubType.ANY_SUBTYPE);
+      return Optional.ofNullable(metadataResult.getEntities()).stream()
+          .flatMap(Collection::stream)
+          .flatMap(
+              result -> {
+                PolarisMetaStoreManager.EntityResult metadataEntityResult =
+                    entityManager
+                        .getMetaStoreManager()
+                        .loadEntity(callContext, result.getCatalogId(), result.getId());
+                return Optional.ofNullable(metadataEntityResult.getEntity())
+                    .map(TableMetadataEntity::of)
+                    .stream();
+              })
+          .filter(
+              metadata -> {
+                if (metadata.getMetadataLocation().equals(metadataLocation)) {
+                  return true;
+                } else {
+                  LOGGER.debug(
+                      String.format("Deleting old entry for %s", metadata.getMetadataLocation()));
+                  entityManager
+                      .getMetaStoreManager()
+                      .dropEntityIfExists(
```

Comment:
Do we have any cleanup for cached metadata entities when the underlying table is dropped?

Reply:
Good callout -- to be consistent with principals, I've added this. In the future I think we need some kind of background cleanup anyway but for now let's clean up when a table is dropped.